server/lib/scheduler/Scheduler.ts

```diff
@@ -122,102 +127,67 @@ class Scheduler {
     }
   }
 
-  public async assignNextWorkspace(userId, maybeSuboptimal = false) {
+  public async assignNextWorkspace(userId: string): Promise<string | null> {
     this.resetCaches();
     this.schedule.leaveCurrentWorkspace(userId);
 
-    let actionableWorkspaces;
-    if (this.isInOracleMode.getValue()) {
-      actionableWorkspaces = await this.getActionableWorkspacesInOracleMode({
-        maybeSuboptimal: false,
-        userId,
-      });
-    } else {
-      actionableWorkspaces = await this.getActionableWorkspaces({
-        maybeSuboptimal: false,
-        userId,
-      });
-    }
-
-    if (actionableWorkspaces.length === 0 && maybeSuboptimal) {
-      if (this.isInOracleMode.getValue()) {
-        actionableWorkspaces = await this.getActionableWorkspacesInOracleMode({
-          maybeSuboptimal: true,
-          userId,
-        });
-      } else {
-        actionableWorkspaces = await this.getActionableWorkspaces({
-          maybeSuboptimal: true,
-          userId,
-        });
-      }
-    }
-
-    if (actionableWorkspaces.length === 0) {
-      const fallbackScheduler = await this.getFallbackScheduler();
-      if (fallbackScheduler) {
-        const assignedWorkspaceId = await fallbackScheduler.assignNextWorkspace(
-          userId,
-          maybeSuboptimal,
-        );
-
-        return assignedWorkspaceId;
-      }
-
-      throw new Error("No eligible workspace");
+    // Should always be in oracle mode
+    if (!this.isInOracleMode.getValue()) {
+      throw new Error(
+        "Not in Oracle Mode. If you see this, please contact us. Then try logging out and logging back in.",
+      );
     }
 
-    const assignedWorkspace = pickRandomItemFromArray(actionableWorkspaces);
-    const isThisAssignmentTimed = await assignedWorkspace.hasTimeBudgetOfRootParent;
+    const actionableWorkspace = await this.getActionableWorkspaceInOracleMode(
+      userId,
+    );
 
-    // This is intended to mitigate a race condition where two users
-    // are both looking for a new assignment at the same time and
-    // get assigned to the same workspace due to async code in the scheduler
-    if (this.schedule.isWorkspaceCurrentlyBeingWorkedOn(assignedWorkspace)) {
```

Comment:
I'd put this back. It's important this check happens as close to the end as possible or else it won't mitigate the race condition aspect of the problem.

Reply:
Yeah fair point. Can't be too safe here, even if it's some extra compute.